Add interpolation of Gauss points to GL points for Bjorhus boundary conditions

nilsdeppe opened the pull request:
## Proposed changes

I don't know how to do Bjorhus boundary conditions when using Gauss points other than to interpolation the boundary time derivative correction to the Gauss points from a Gauss-Lobatto grid. This PR adds the interpolation functions and matrix calculations.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add interpolation for Gauss points into the volume

This is needed for Bjorhus boundary conditions where we need to interpolate
du/dt back into the volume

nilsdeppe committed:
Add function to interpolate Bjorhus corrections to Gauss points

kidder commented on src/Evolution/DiscontinuousGalerkin/InterpolateFromBoundary.hpp:
```diff
@@ -0,0 +1,72 @@
+#include <utility>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
```
is this include needed?

kidder commented on src/Evolution/DiscontinuousGalerkin/InterpolateFromBoundary.cpp:
```diff
@@ -0,0 +1,99 @@
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/StripeIterator.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
```
not needed

kidder commented on tests/Unit/Evolution/DiscontinuousGalerkin/Test_InterpolateFromBoundary.cpp:
```diff
@@ -0,0 +1,93 @@
+#include "Domain/LogicalCoordinates.hpp"
+#include "Domain/Structure/Direction.hpp"
+#include "Evolution/DiscontinuousGalerkin/InterpolateFromBoundary.hpp"
+#include "NumericalAlgorithms/Interpolation/IrregularInterpolant.hpp"
```
unused

kidder commented on tests/Unit/Evolution/DiscontinuousGalerkin/Test_InterpolateFromBoundary.cpp:
```diff
@@ -0,0 +1,93 @@
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/SliceIterator.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/LogicalCoordinates.hpp"
```
unused

kidder reviewed (changes requested):
you can squash these

nilsdeppe commented:
I've rebased and squashed. Thanks for the review :)

fmahebert commented on src/NumericalAlgorithms/Spectral/Spectral.hpp:
```diff
@@ -445,6 +445,31 @@ const std::pair<Matrix, Matrix>& boundary_interpolation_matrices(
+/*!
+ * \brief Interpolates values from the boundary into the volume, which is needed
+ * when applying time derivative or Bjorhus-type boundary conditions in a
+ * discontinuous Galerkin scheme.
```
1. Is it true that this is only needed when evolving on Gauss points? If so, I would add that clarification here, perhaps: "... in a discontinuous Galerkin scheme using Gauss points." or similar.

2. I would also add a new paragraph (or perhaps expand the paragraph below?) where you clarify what "interpolating from boundary into volume" means, because it's not obvious and didn't make sense to me at first. I think it would help to say that you start from data on the boundary, you interpret this data as being a Lagrange polynomial with the correct amplitude on the bdry and value of 0 at every other LGL grid point, then you interpolate from this Lagrange polynomial onto the set of Gauss points. This is maybe slightly too much detail, but should hopefully clear up any confusion. Thoughts?

3. I think it would also be nice to add another new (short) paragraph clarifying how this is different from lifting. As far as I can tell, it's the same idea (boundary => volume) but with different amplitudes?

fmahebert commented on src/NumericalAlgorithms/Spectral/Spectral.hpp:
```diff
@@ -445,6 +445,31 @@ const std::pair<Matrix, Matrix>& boundary_interpolation_matrices(
+ * \f{align*}{
+ *   &\ell^{\mathrm{Gauss-Lobatto}}_{0}(\xi_j^{\mathrm{Gauss}}) \\
+ *   &\ell^{\mathrm{Gauss-Lobatto}}_{N}(\xi_j^{\mathrm{Gauss}})
+ * \f}
```
I did not find this math helpful... there's no `=`, and I don't know what is being described... what's going on?

fmahebert commented on src/NumericalAlgorithms/Spectral/Spectral.cpp:
```diff
@@ -405,6 +405,47 @@ const std::pair<Matrix, Matrix>& boundary_interpolation_matrices(
+  static_assert(BasisType == Spectral::Basis::Legendre);
+  static_assert(
+      QuadratureType == Spectral::Quadrature::Gauss,
+      "We only compute the boundary lifting for Gauss quadrature "
```
Update this comment (this is not the lifting term)

fmahebert commented on src/NumericalAlgorithms/Spectral/Spectral.cpp:
```diff
@@ -405,6 +405,47 @@ const std::pair<Matrix, Matrix>& boundary_interpolation_matrices(
+      [](const size_t local_num_points) noexcept {
+        const Matrix interp_matrix =
+            interpolation_matrix<BasisType, Quadrature::GaussLobatto>(
+                local_num_points == 1 ? local_num_points + 1 : local_num_points,
```
trivial, but I'd suggest `local_num_points + 1` => `2` which is easier to parse

fmahebert commented on src/NumericalAlgorithms/Spectral/Spectral.cpp:
```diff
@@ -405,6 +405,47 @@ const std::pair<Matrix, Matrix>& boundary_interpolation_matrices(
+  ASSERT(mesh.basis(0) == Spectral::Basis::Legendre,
+         "We only support DG with a Legendre basis.");
+  ASSERT(mesh.quadrature(0) == Spectral::Quadrature::Gauss,
+         "We only compute the boundary lifting for Gauss quadrature "
```
Update this comment

nilsdeppe commented:
I've rebased on develop and pushed a fixup that hopefully addresses the comments. I reworked the documentation a decent amount, so it didn't strictly follow the suggestions, but hopefully the correct information is now available :)

fmahebert commented on src/NumericalAlgorithms/Spectral/Spectral.hpp:
```diff
@@ -449,17 +449,26 @@ const std::pair<Matrix, Matrix>& boundary_interpolation_matrices(
+ * \f$\ell^{\mathrm{Gauss-Lobatto}}_{N}\f$ polynomials/basis functions
+ * contribute to the correction. In order to interpolate the correction from the
+ * nodes at the boundary, the Gauss-Lobatto Lagrange polynomials  must be
+ * evaluated an the Gauss grid points. The returned pair of `DataVector`s stores
```
"an" -> "at"

fmahebert reviewed (commented):
I think the new docs are well improved. Thanks!

You have my permission to squash, including this typo

nilsdeppe commented:
Squashed. Thanks for the reviews :)
